File: tools/memory/compress.ts
#!/usr/bin/env npx tsx

import { createHash } from "crypto";
import { runPsql } from "../lib/db.js";

type EpisodicEntry = {
  id: number;
  happened_at: string;
  summary: string;
  details: string;
  tags: string[];
  participants: string[];
};

type Cluster = {
  id: number;
  entries: EpisodicEntry[];
  featureCounts: Map<string, number>;
};

const GENERIC_TAGS = new Set([
  "memory",
  "daily_memory",
  "heartbeat_ingest",
  "note",
  "notes",
  "log",
  "general",
]);

const STOPWORDS = new Set([
  "the",
  "and",
  "for",
  "that",
  "with",
  "from",
  "this",
  "have",
  "will",
  "were",
  "been",
  "about",
  "into",
  "when",
  "what",
  "where",
  "your",
  "their",
  "them",
  "then",
  "than",
  "just",
  "also",
  "more",
  "very",
  "only",
  "over",
  "under",
  "need",
  "next",
]);

function q(value: unknown): string {
  if (value === null || value === undefined) return "NULL";
  return `'${String(value).replace(/'/g, "''")}'`;
}

function psql(sql: string, capture = false): string {
  const res = runPsql(sql, {
    args: ["-q", "-v", "ON_ERROR_STOP=1", "-A", "-t", "-F", "\t", "-X"],
  });
  if (res.status !== 0) {
    const msg = (res.stderr ?? "").trim() || "psql command failed";
    throw new Error(msg);
  }
  if (!capture) return "";
  return (res.stdout ?? "").trim();
}

function fp(...parts: string[]): string {
  const h = createHash("sha256");
  for (const p of parts) {
    h.update(p ?? "");
    h.update("|");
  }
  return h.digest("hex").slice(0, 40);
}

function ensureFidelityColumn(): void {
  const sql = `
    ALTER TABLE cortana_memory_semantic
    ADD COLUMN IF NOT EXISTS fidelity_score numeric(5,4);

    DO $$
    BEGIN
      IF NOT EXISTS (
        SELECT 1
        FROM pg_constraint
        WHERE conname = 'cortana_memory_semantic_fidelity_score_check'
      ) THEN
        ALTER TABLE cortana_memory_semantic
        ADD CONSTRAINT cortana_memory_semantic_fidelity_score_check
        CHECK (fidelity_score IS NULL OR (fidelity_score >= 0 AND fidelity_score <= 1));
      END IF;
    END $$;
    `;
  psql(sql);
}

function parseArray(raw: string): string[] {
  if (!raw) return [];
  return raw.split(",").filter((x) => x);
}

function fetchRecentEpisodic(sinceHours: number): EpisodicEntry[] {
  const sql = `
    SELECT
      id,
      happened_at::text,
      COALESCE(summary, ''),
      COALESCE(details, ''),
      COALESCE(array_to_string(tags, ','), ''),
      COALESCE(array_to_string(participants, ','), '')
    FROM cortana_memory_episodic
    WHERE active = TRUE
      AND happened_at >= NOW() - INTERVAL '${sinceHours} hours'
    ORDER BY happened_at ASC;
    `;
  const out = psql(sql, true);
  const rows: EpisodicEntry[] = [];
  for (const line of out.split(/\r?\n/)) {
    if (!line.trim()) continue;
    const parts = line.split("\t");
    if (parts.length !== 6) continue;
    rows.push({
      id: Number.parseInt(parts[0] ?? "0", 10),
      happened_at: parts[1] ?? "",
      summary: parts[2] ?? "",
      details: parts[3] ?? "",
      tags: parseArray(parts[4] ?? ""),
      participants: parseArray(parts[5] ?? ""),
    });
  }
  return rows;
}

function tokenize(text: string): string[] {
  const words = text.match(/[A-Za-z][A-Za-z0-9_-]{2,}/g) ?? [];
  return words.map((w) => w.toLowerCase()).filter((w) => !STOPWORDS.has(w));
}

function extractFeatures(entry: EpisodicEntry): Set<string> {
  const tags = new Set(
    entry.tags
      .map((t) => t.trim().toLowerCase())
      .filter((t) => t && !GENERIC_TAGS.has(t))
  );

  const toks = tokenize(`${entry.summary}\n${entry.details}`);
  const counts = new Map<string, number>();
  for (const t of toks) {
    counts.set(t, (counts.get(t) ?? 0) + 1);
  }
  const common = Array.from(counts.entries())
    .sort((a, b) => b[1] - a[1])
    .slice(0, 8)
    .map(([w]) => w);

  for (const w of common) tags.add(w);
  return tags;
}

function jaccard(a: Set<string>, b: Set<string>): number {
  if (!a.size || !b.size) return 0.0;
  let inter = 0;
  for (const v of a) {
    if (b.has(v)) inter += 1;
  }
  const union = a.size + b.size - inter;
  return union ? inter / union : 0.0;
}

function clusterEntries(entries: EpisodicEntry[], threshold = 0.25): Cluster[] {
  const clusters: Cluster[] = [];
  for (const entry of entries) {
    const features = extractFeatures(entry);
    let bestIdx = -1;
    let bestScore = 0.0;
    for (let i = 0; i < clusters.length; i += 1) {
      const c = clusters[i];
      const featureSet = new Set(c.featureCounts.keys());
      const score = jaccard(features, featureSet);
      if (score > bestScore) {
        bestScore = score;
        bestIdx = i;
      }
    }

    let target: Cluster;
    if (bestIdx >= 0 && bestScore >= threshold) {
      target = clusters[bestIdx];
    } else {
      target = { id: clusters.length + 1, entries: [], featureCounts: new Map() };
      clusters.push(target);
    }

    target.entries.push(entry);
    for (const f of features) {
      target.featureCounts.set(f, (target.featureCounts.get(f) ?? 0) + 1);
    }
  }
  return clusters;
}

function normalizeLine(line: string): string {
  return line.replace(/\s+/g, " ").trim().replace(/^[-•\t ]+|[-•\t ]+$/g, "");
}

function pickItems(lines: string[], patterns: string[], limit = 6): string[] {
  const out: string[] = [];
  const seen = new Set<string>();
  const regexes = patterns.map((p) => new RegExp(p, "i"));
  for (const raw of lines) {
    const l = normalizeLine(raw);
    if (!l) continue;
    if (regexes.some((r) => r.test(l))) {
      const key = l.toLowerCase();
      if (seen.has(key)) continue;
      seen.add(key);
      out.push(l);
      if (out.length >= limit) break;
    }
  }
  return out;
}

function extractEntities(text: string): Set<string> {
  const entities = new Set<string>();
  const caps = text.match(/\b[A-Z][a-zA-Z0-9_+-]{2,}\b/g) ?? [];
  for (const m of caps) entities.add(m.toLowerCase());
  const ids = text.match(/\b(?:task|issue|pr|ticket)[\s:#-]*\d+\b/gi) ?? [];
  for (const m of ids) entities.add(m.toLowerCase().replace(/\s+/g, ""));
  const nums = text.match(/\b\d+(?:\.\d+)?(?:%|h|hr|hrs|am|pm|days?|weeks?)\b/gi) ?? [];
  for (const m of nums) entities.add(m.toLowerCase());
  return entities;
}

function buildCompression(cluster: Cluster): Record<string, any> {
  const allLines: string[] = [];
  const combinedTextParts: string[] = [];

  for (const e of cluster.entries) {
    const combined = `${e.summary || ""}\n${e.details || ""}`.trim();
    combinedTextParts.push(combined);
    const parts = combined.split(/[\n\.]+/);
    for (const p of parts) {
      if (p && p.trim()) allLines.push(p);
    }
  }

  const facts = pickItems(allLines, ["\\b(is|are|was|were|has|have|had|updated|completed|deployed|fixed)\\b", "\\b\\d"], 7);
  const decisions = pickItems(allLines, ["\\b(decid|chose|choice|plan|will|should|recommend|approved)\\b"], 5);
  const actions = pickItems(allLines, ["\\b(todo|action|follow up|next step|need to|pending|blocker|deadline)\\b"], 6);

  const topTopics = Array.from(cluster.featureCounts.entries())
    .sort((a, b) => b[1] - a[1])
    .slice(0, 4)
    .map(([k]) => k);
  const topicLabel = topTopics.length ? topTopics.join(", ") : `cluster-${cluster.id}`;

  const bodyLines: string[] = [`Topic: ${topicLabel}`];
  if (facts.length) {
    bodyLines.push("Key facts:");
    bodyLines.push(...facts.map((x) => `- ${x}`));
  }
  if (decisions.length) {
    bodyLines.push("Decisions:");
    bodyLines.push(...decisions.map((x) => `- ${x}`));
  }
  if (actions.length) {
    bodyLines.push("Action items:");
    bodyLines.push(...actions.map((x) => `- ${x}`));
  }

  if (!facts.length && !decisions.length && !actions.length) {
    const fallback = allLines
      .map((x) => normalizeLine(x))
      .filter((x) => x)
      .slice(0, 8);
    if (fallback.length) {
      bodyLines.push("Summary:");
      bodyLines.push(...fallback.map((x) => `- ${x}`));
    }
  }

  const compressedText = bodyLines.join("\n").trim();
  const sourceText = combinedTextParts.join("\n");

  const sourceEntities = extractEntities(sourceText);
  const compressedEntities = extractEntities(compressedText);
  const overlap = new Set<string>();
  for (const e of sourceEntities) {
    if (compressedEntities.has(e)) overlap.add(e);
  }

  let fidelity = 0.0;
  if (sourceEntities.size) {
    fidelity = overlap.size / sourceEntities.size;
  } else {
    fidelity = compressedText ? 1.0 : 0.0;
  }

  const srcHasDecisions = /\b(decid|chose|approved|plan)\b/i.test(sourceText);
  const srcHasActions = /\b(todo|follow up|next step|need to|pending|blocker)\b/i.test(sourceText);
  if (srcHasDecisions && !decisions.length) fidelity *= 0.9;
  if (srcHasActions && !actions.length) fidelity *= 0.9;

  fidelity = Math.max(0.0, Math.min(1.0, Math.round(fidelity * 10000) / 10000));

  return {
    topic: topicLabel,
    text: compressedText,
    facts,
    decisions,
    actions,
    fidelity,
    source_entities: Array.from(sourceEntities).sort(),
    compressed_entities: Array.from(compressedEntities).sort(),
    entity_overlap: Array.from(overlap).sort(),
  };
}

function insertSemantic(
  compression: Record<string, any>,
  cluster: Cluster,
  sinceHours: number,
  dryRun: boolean
): boolean {
  const entryIds = cluster.entries.map((e) => e.id);
  const subject = `memory_cluster:${cluster.id}`;
  const predicate = "daily_distillation";
  const objectValue = compression.text;
  const sourceRef = `episodic:${Math.min(...entryIds)}-${Math.max(...entryIds)}:h${sinceHours}`;
  const fingerprint = fp(subject, predicate, objectValue, sourceRef);

  const metadata = {
    engine: "semantic-compression-v1",
    entry_ids: entryIds,
    entry_count: entryIds.length,
    topic: compression.topic,
    facts: compression.facts,
    decisions: compression.decisions,
    actions: compression.actions,
    source_entities: compression.source_entities,
    compressed_entities: compression.compressed_entities,
    entity_overlap: compression.entity_overlap,
  };

  const confidence = Math.max(0.5, Math.min(0.98, 0.55 + compression.fidelity * 0.4));

  const sql = `
    INSERT INTO cortana_memory_semantic
      (fact_type, subject, predicate, object_value, confidence, trust, stability,
       first_seen_at, last_seen_at, source_type, source_ref, fingerprint, metadata, fidelity_score)
    VALUES
      ('fact', ${q(subject)}, ${q(predicate)}, ${q(objectValue)},
       ${confidence.toFixed(4)}, 0.850, 0.600, NOW(), NOW(),
       'semantic_compression', ${q(sourceRef)}, ${q(fingerprint)}, ${q(JSON.stringify(metadata))}::jsonb,
       ${compression.fidelity.toFixed(4)})
    ON CONFLICT (fact_type, subject, predicate, object_value)
    DO UPDATE SET
      last_seen_at = NOW(),
      source_type = EXCLUDED.source_type,
      source_ref = EXCLUDED.source_ref,
      fingerprint = EXCLUDED.fingerprint,
      metadata = EXCLUDED.metadata,
      confidence = EXCLUDED.confidence,
      fidelity_score = EXCLUDED.fidelity_score,
      active = TRUE;
    `;

  if (dryRun) return true;
  psql(sql);
  return true;
}

function printHelp(): void {
  const text = `usage: compress.ts [-h] [--since-hours SINCE_HOURS] [--dry-run] [--min-cluster-size MIN_CLUSTER_SIZE]\n\nCompress recent episodic memories into semantic memory.\n\noptions:\n  -h, --help            show this help message and exit\n  --since-hours SINCE_HOURS  Lookback window in hours (recommended 24-48).\n  --dry-run             Analyze without writing to DB.\n  --min-cluster-size MIN_CLUSTER_SIZE  Skip clusters smaller than this size.`;
  console.log(text);
}

function parseArgs(argv: string[]): { sinceHours: number; dryRun: boolean; minClusterSize: number } {
  const args = { sinceHours: 36, dryRun: false, minClusterSize: 1 };
  for (let i = 0; i < argv.length; i += 1) {
    const arg = argv[i];
    const next = argv[i + 1];
    if (arg === "-h" || arg === "--help") {
      printHelp();
      process.exit(0);
    } else if (arg === "--since-hours" && next) {
      args.sinceHours = Number.parseInt(next, 10);
      i += 1;
    } else if (arg === "--dry-run") {
      args.dryRun = true;
    } else if (arg === "--min-cluster-size" && next) {
      args.minClusterSize = Number.parseInt(next, 10);
      i += 1;
    } else if (arg.startsWith("-")) {
      console.error(`Unknown argument: ${arg}`);
      printHelp();
      process.exit(2);
    }
  }
  return args;
}

async function main(): Promise<void> {
  const args = parseArgs(process.argv.slice(2));
  const sinceHours = Math.max(1, Math.min(168, args.sinceHours));

  ensureFidelityColumn();
  const rows = fetchRecentEpisodic(sinceHours);
  if (!rows.length) {
    console.log(
      JSON.stringify({ ok: true, message: "No episodic memories in lookback window.", since_hours: sinceHours })
    );
    return;
  }

  const clusters = clusterEntries(rows);
  let written = 0;
  const outputs: Array<Record<string, any>> = [];

  for (const c of clusters) {
    if (c.entries.length < args.minClusterSize) continue;
    const comp = buildCompression(c);
    insertSemantic(comp, c, sinceHours, args.dryRun);
    written += 1;
    outputs.push({
      cluster_id: c.id,
      entry_ids: c.entries.map((e) => e.id),
      entry_count: c.entries.length,
      topic: comp.topic,
      fidelity: comp.fidelity,
    });
  }

  const out = {
    ok: true,
    dry_run: args.dryRun,
    since_hours: sinceHours,
    episodic_entries: rows.length,
    clusters_total: clusters.length,
    clusters_written: written,
    results: outputs,
  };
  console.log(JSON.stringify(out, null, 2));
}

main().catch((err) => {
  const msg = err instanceof Error ? err.message : String(err);
  console.error(msg);
  process.exit(1);
});

File: tools/memory/compress.py
#!/usr/bin/env python3
"""Semantic Compression Engine: Daily Memory Distillation with Fidelity Checks.

Reads recent episodic memories, clusters related entries, generates compressed
summaries (facts/decisions/action items), computes fidelity scores, and writes
results to cortana_memory_semantic.
"""

from __future__ import annotations

import argparse
import hashlib
import json
import re
import subprocess
from collections import Counter
from dataclasses import dataclass, field
from datetime import datetime, timedelta, timezone
from pathlib import Path
from typing import Any

PSQL = "/opt/homebrew/opt/postgresql@17/bin/psql"
DB = "cortana"
WORKSPACE = Path("/Users/hd/openclaw")

GENERIC_TAGS = {
    "memory",
    "daily_memory",
    "heartbeat_ingest",
    "note",
    "notes",
    "log",
    "general",
}
STOPWORDS = {
    "the",
    "and",
    "for",
    "that",
    "with",
    "from",
    "this",
    "have",
    "will",
    "were",
    "been",
    "about",
    "into",
    "when",
    "what",
    "where",
    "your",
    "their",
    "them",
    "then",
    "than",
    "just",
    "also",
    "more",
    "very",
    "only",
    "over",
    "under",
    "need",
    "next",
}


@dataclass
class EpisodicEntry:
    id: int
    happened_at: str
    summary: str
    details: str
    tags: list[str]
    participants: list[str]

    @property
    def text(self) -> str:
        return "\n".join([self.summary or "", self.details or ""]).strip()


@dataclass
class Cluster:
    id: int
    entries: list[EpisodicEntry] = field(default_factory=list)
    feature_counts: Counter = field(default_factory=Counter)

    @property
    def feature_set(self) -> set[str]:
        return set(self.feature_counts.keys())


def q(value: Any) -> str:
    if value is None:
        return "NULL"
    return "'" + str(value).replace("'", "''") + "'"


def psql(sql: str, capture: bool = False) -> str:
    proc = subprocess.run(
        [PSQL, DB, "-q", "-v", "ON_ERROR_STOP=1", "-At", "-F", "\t", "-c", sql],
        capture_output=True,
        text=True,
    )
    if proc.returncode != 0:
        raise RuntimeError(proc.stderr.strip() or "psql command failed")
    if not capture:
        return ""
    return (proc.stdout or "").strip()


def fp(*parts: str) -> str:
    h = hashlib.sha256()
    for p in parts:
        h.update((p or "").encode("utf-8"))
        h.update(b"|")
    return h.hexdigest()[:40]


def ensure_fidelity_column() -> None:
    sql = """
    ALTER TABLE cortana_memory_semantic
    ADD COLUMN IF NOT EXISTS fidelity_score numeric(5,4);

    DO $$
    BEGIN
      IF NOT EXISTS (
        SELECT 1
        FROM pg_constraint
        WHERE conname = 'cortana_memory_semantic_fidelity_score_check'
      ) THEN
        ALTER TABLE cortana_memory_semantic
        ADD CONSTRAINT cortana_memory_semantic_fidelity_score_check
        CHECK (fidelity_score IS NULL OR (fidelity_score >= 0 AND fidelity_score <= 1));
      END IF;
    END $$;
    """
    psql(sql)


def parse_array(raw: str) -> list[str]:
    if not raw:
        return []
    return [x for x in raw.split(",") if x]


def fetch_recent_episodic(since_hours: int) -> list[EpisodicEntry]:
    sql = f"""
    SELECT
      id,
      happened_at::text,
      COALESCE(summary, ''),
      COALESCE(details, ''),
      COALESCE(array_to_string(tags, ','), ''),
      COALESCE(array_to_string(participants, ','), '')
    FROM cortana_memory_episodic
    WHERE active = TRUE
      AND happened_at >= NOW() - INTERVAL '{since_hours} hours'
    ORDER BY happened_at ASC;
    """
    out = psql(sql, capture=True)
    rows: list[EpisodicEntry] = []
    for line in out.splitlines():
        if not line.strip():
            continue
        parts = line.split("\t")
        if len(parts) != 6:
            continue
        rows.append(
            EpisodicEntry(
                id=int(parts[0]),
                happened_at=parts[1],
                summary=parts[2],
                details=parts[3],
                tags=parse_array(parts[4]),
                participants=parse_array(parts[5]),
            )
        )
    return rows


def tokenize(text: str) -> list[str]:
    words = re.findall(r"[A-Za-z][A-Za-z0-9_-]{2,}", text.lower())
    return [w for w in words if w not in STOPWORDS]


def extract_features(entry: EpisodicEntry) -> set[str]:
    tags = {t.strip().lower() for t in entry.tags if t and t.strip()}
    tags = {t for t in tags if t not in GENERIC_TAGS}

    toks = tokenize(entry.summary + "\n" + entry.details)
    common = {w for w, _ in Counter(toks).most_common(8)}
    return tags.union(common)


def jaccard(a: set[str], b: set[str]) -> float:
    if not a or not b:
        return 0.0
    inter = len(a & b)
    union = len(a | b)
    return inter / union if union else 0.0


def cluster_entries(entries: list[EpisodicEntry], threshold: float = 0.25) -> list[Cluster]:
    clusters: list[Cluster] = []
    for entry in entries:
        features = extract_features(entry)
        best_idx = -1
        best_score = 0.0
        for i, c in enumerate(clusters):
            score = jaccard(features, c.feature_set)
            if score > best_score:
                best_score = score
                best_idx = i

        if best_idx >= 0 and best_score >= threshold:
            target = clusters[best_idx]
        else:
            target = Cluster(id=len(clusters) + 1)
            clusters.append(target)

        target.entries.append(entry)
        for f in features:
            target.feature_counts[f] += 1
    return clusters


def normalize_line(line: str) -> str:
    line = re.sub(r"\s+", " ", line).strip(" -•\t")
    return line


def pick_items(lines: list[str], patterns: list[str], limit: int = 6) -> list[str]:
    out: list[str] = []
    seen: set[str] = set()
    for raw in lines:
        l = normalize_line(raw)
        if not l:
            continue
        if any(re.search(p, l, flags=re.IGNORECASE) for p in patterns):
            key = l.lower()
            if key in seen:
                continue
            seen.add(key)
            out.append(l)
            if len(out) >= limit:
                break
    return out


def extract_entities(text: str) -> set[str]:
    entities: set[str] = set()
    # Capitalized words / names / products.
    for m in re.findall(r"\b[A-Z][a-zA-Z0-9_+-]{2,}\b", text):
        entities.add(m.lower())
    # IDs and ticket/task references.
    for m in re.findall(r"\b(?:task|issue|pr|ticket)[\s:#-]*\d+\b", text, flags=re.IGNORECASE):
        entities.add(m.lower().replace(" ", ""))
    # Dates/numbers with units.
    for m in re.findall(r"\b\d+(?:\.\d+)?(?:%|h|hr|hrs|am|pm|days?|weeks?)\b", text, flags=re.IGNORECASE):
        entities.add(m.lower())
    return entities


def build_compression(cluster: Cluster) -> dict[str, Any]:
    all_lines: list[str] = []
    combined_text_parts: list[str] = []

    for e in cluster.entries:
        combined = e.text
        combined_text_parts.append(combined)
        parts = re.split(r"[\n\.]+", combined)
        all_lines.extend([p for p in parts if p and p.strip()])

    facts = pick_items(
        all_lines,
        patterns=[r"\b(is|are|was|were|has|have|had|updated|completed|deployed|fixed)\b", r"\b\d"],
        limit=7,
    )
    decisions = pick_items(
        all_lines,
        patterns=[r"\b(decid|chose|choice|plan|will|should|recommend|approved)\b"],
        limit=5,
    )
    actions = pick_items(
        all_lines,
        patterns=[r"\b(todo|action|follow up|next step|need to|pending|blocker|deadline)\b"],
        limit=6,
    )

    top_topics = [k for k, _ in cluster.feature_counts.most_common(4)]
    topic_label = ", ".join(top_topics) if top_topics else f"cluster-{cluster.id}"

    body_lines = [f"Topic: {topic_label}"]
    if facts:
        body_lines.append("Key facts:")
        body_lines.extend([f"- {x}" for x in facts])
    if decisions:
        body_lines.append("Decisions:")
        body_lines.extend([f"- {x}" for x in decisions])
    if actions:
        body_lines.append("Action items:")
        body_lines.extend([f"- {x}" for x in actions])

    if not facts and not decisions and not actions:
        fallback = [normalize_line(x) for x in all_lines[:8] if normalize_line(x)]
        if fallback:
            body_lines.append("Summary:")
            body_lines.extend([f"- {x}" for x in fallback])

    compressed_text = "\n".join(body_lines).strip()
    source_text = "\n".join(combined_text_parts)

    source_entities = extract_entities(source_text)
    compressed_entities = extract_entities(compressed_text)
    overlap = source_entities & compressed_entities

    if source_entities:
        fidelity = len(overlap) / len(source_entities)
    else:
        fidelity = 1.0 if compressed_text else 0.0

    # Penalize if we had explicit actions/decisions in source but missed all.
    src_has_decisions = bool(re.search(r"\b(decid|chose|approved|plan)\b", source_text, flags=re.IGNORECASE))
    src_has_actions = bool(re.search(r"\b(todo|follow up|next step|need to|pending|blocker)\b", source_text, flags=re.IGNORECASE))
    if src_has_decisions and not decisions:
        fidelity *= 0.9
    if src_has_actions and not actions:
        fidelity *= 0.9

    fidelity = max(0.0, min(1.0, round(fidelity, 4)))

    return {
        "topic": topic_label,
        "text": compressed_text,
        "facts": facts,
        "decisions": decisions,
        "actions": actions,
        "fidelity": fidelity,
        "source_entities": sorted(source_entities),
        "compressed_entities": sorted(compressed_entities),
        "entity_overlap": sorted(overlap),
    }


def insert_semantic(
    compression: dict[str, Any],
    cluster: Cluster,
    since_hours: int,
    dry_run: bool,
) -> bool:
    entry_ids = [e.id for e in cluster.entries]
    subject = f"memory_cluster:{cluster.id}"
    predicate = "daily_distillation"
    object_value = compression["text"]
    source_ref = f"episodic:{min(entry_ids)}-{max(entry_ids)}:h{since_hours}"
    fingerprint = fp(subject, predicate, object_value, source_ref)

    metadata = {
        "engine": "semantic-compression-v1",
        "entry_ids": entry_ids,
        "entry_count": len(entry_ids),
        "topic": compression["topic"],
        "facts": compression["facts"],
        "decisions": compression["decisions"],
        "actions": compression["actions"],
        "source_entities": compression["source_entities"],
        "compressed_entities": compression["compressed_entities"],
        "entity_overlap": compression["entity_overlap"],
    }

    confidence = max(0.50, min(0.98, 0.55 + compression["fidelity"] * 0.4))

    sql = f"""
    INSERT INTO cortana_memory_semantic
      (fact_type, subject, predicate, object_value, confidence, trust, stability,
       first_seen_at, last_seen_at, source_type, source_ref, fingerprint, metadata, fidelity_score)
    VALUES
      ('fact', {q(subject)}, {q(predicate)}, {q(object_value)},
       {confidence:.4f}, 0.850, 0.600, NOW(), NOW(),
       'semantic_compression', {q(source_ref)}, {q(fingerprint)}, {q(json.dumps(metadata))}::jsonb,
       {compression['fidelity']:.4f})
    ON CONFLICT (fact_type, subject, predicate, object_value)
    DO UPDATE SET
      last_seen_at = NOW(),
      source_type = EXCLUDED.source_type,
      source_ref = EXCLUDED.source_ref,
      fingerprint = EXCLUDED.fingerprint,
      metadata = EXCLUDED.metadata,
      confidence = EXCLUDED.confidence,
      fidelity_score = EXCLUDED.fidelity_score,
      active = TRUE;
    """

    if dry_run:
        return True
    psql(sql)
    return True


def main() -> None:
    ap = argparse.ArgumentParser(description="Compress recent episodic memories into semantic memory.")
    ap.add_argument("--since-hours", type=int, default=36, help="Lookback window in hours (recommended 24-48).")
    ap.add_argument("--dry-run", action="store_true", help="Analyze without writing to DB.")
    ap.add_argument("--min-cluster-size", type=int, default=1, help="Skip clusters smaller than this size.")
    args = ap.parse_args()

    since_hours = max(1, min(168, args.since_hours))

    ensure_fidelity_column()
    rows = fetch_recent_episodic(since_hours)
    if not rows:
        print(json.dumps({"ok": True, "message": "No episodic memories in lookback window.", "since_hours": since_hours}))
        return

    clusters = cluster_entries(rows)
    written = 0
    outputs: list[dict[str, Any]] = []

    for c in clusters:
        if len(c.entries) < args.min_cluster_size:
            continue
        comp = build_compression(c)
        insert_semantic(comp, c, since_hours=since_hours, dry_run=args.dry_run)
        written += 1
        outputs.append(
            {
                "cluster_id": c.id,
                "entry_ids": [e.id for e in c.entries],
                "entry_count": len(c.entries),
                "topic": comp["topic"],
                "fidelity": comp["fidelity"],
            }
        )

    out = {
        "ok": True,
        "dry_run": args.dry_run,
        "since_hours": since_hours,
        "episodic_entries": len(rows),
        "clusters_total": len(clusters),
        "clusters_written": written,
        "results": outputs,
    }
    print(json.dumps(out, indent=2))


if __name__ == "__main__":
    main()
